src/diffusion/model.py

```diff
@@ -64,3 +88,92 @@ def forward(
         ).last_hidden_state
 
         return output
+
+    def training_step(self, batch, batch_idx):
+        context, gt = batch
+
+        emb = self.emb(gt)
+
+        x_0, x_t, t = get_diffusion_variables(self.diffusion_steps, emb, self.alphas_cumprod_prev)
+
+        x_0_hat = self(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_t,
+            time_ids=t,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+        probs = self.classifier(x_0)
+        ce = calculate_batch_ce(probs, gt, pad_mask)
+
+        emb_mask = ~pad_mask.unsqueeze(-1)
+
+        mse = torch.where(
+            t == 1,
+            flat_mean((x_0_hat - emb) ** 2 * emb_mask),
+            flat_mean((x_0_hat - x_0) ** 2 * emb_mask),
+        ).mean()
+
+        t0_loss = (x_0**2 * emb_mask).mean()
+
+        loss = mse + ce + t0_loss
+
+        metrics = {
+            "train/mse": mse.item(),
+            "train/ce": ce.item(),
+            "train/t0": t0_loss.item(),
+            "train/loss": loss.item(),
+        }
+
+        self.log_dict(metrics)
+
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        context, gt = batch
+
+        emb = self.emb(gt)
+        x_0 = prepare_x0(emb)
+
+        ones = torch.ones(x_0.shape[0]).long().to(x_0.device)
+
+        for time in range(self.diffusion_steps, 1, -self.step_freq):
+            t = ones * time
+            x_t = get_xt(x_0, self.alphas_cumprod_prev, t)
+
+            x_0 = self(
+                encoder_input_ids=context,
+                decoder_inputs_embeds=x_t,
+                time_ids=t,
+            )
+
+        x_1 = get_xt(x_0, self.alphas_cumprod_prev, ones)
+
+        x_0 = self(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_1,
+            time_ids=ones,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+        probs = self.classifier(x_0)
+        ce = calculate_batch_ce(probs, gt, pad_mask)
+
+        emb_mask = ~pad_mask
+        preds = probs.argmax(-1)
+
+        acc = (((preds == gt) * emb_mask).sum(axis=1) / emb_mask.sum(axis=1)).mean()
+
+        return torch.tensor([ce.item(), acc.item()])
```

Comment:
No need to create tensor here and call `.item()`. You can use `.detach()`, but I don't think that this would speed up. Return here just tuple: `return ce, acc`